tools/c7n_mailer/c7n_mailer/slack_delivery.py

```diff
@@ -0,0 +1,134 @@
+                    slack_messages[address] = get_rendered_jinja(
+                        slack_target, sqs_message, to_addrs_to_resources_map.values()[0],
+                        self.logger, 'slack_template', 'slack_default')
+                self.logger.debug("Sending to specified email target, based on lookup via Slack API.")
+            elif e.startswith('slack://#'):
+                resolved_addrs = e.split('slack://#', 1)[1]
+                slack_messages[resolved_addrs] = get_rendered_jinja(resolved_addrs, sqs_message,
+                                                                    to_addrs_to_resources_map.values()[0],
+                                                                    self.logger, 'slack_template', 'slack_default')
+
+                self.logger.debug("Sending to specified Slack channel.")
+
+        return slack_messages
+
+    def slack_handler(self, sqs_message, slack_messages):
+        for key, payload in slack_messages.iteritems():
+            self.logger.info("Sending account:%s policy:%s %s:%s slack:%s to %s" % (
+                sqs_message.get('account', ''),
+                sqs_message['policy']['name'],
+                sqs_message['policy']['resource'],
+                str(len(sqs_message['resources'])),
+                sqs_message['action'].get('slack_template', 'slack_default'),
+                json.loads(payload)["channel"])
+            )
+
+            self.send_slack_msg(payload)
+
+    def retrieve_user_im(self, email_addresses):
+        list = {}
+
+        for address in email_addresses:
+            if self.cache_engine and self.caching.get(address):
+                self.logger.debug('Got slack metadata from local cache for: %s' % address)
+                list[address] = self.caching.get(address)
+                continue
+
+            response = self.client.api_call(
+                "users.lookupByEmail", email=address)
+
+            if not response["ok"] and "Retry-After" in response["headers"]:
+                self.logger.info("Slack API rate limiting. Waiting %d seconds") % (int(response.headers['retry-after']))
+                time.sleep(int(response.headers['Retry-After']))
+                continue
+            elif not response["ok"] and response["error"] == "invalid_auth":
+                raise Exception("Invalid Slack token.")
+            elif not response["ok"] and response["error"] == "users_not_found":
+                self.logger.info("Slack user ID not found.")
+                self.caching.set(address, {})
+                continue
+            else:
+                self.logger.debug("Slack account %s found for user %s", response['user']['enterprise_user']['id'])
+                if self.cache_engine:
+                    self.logger.debug('Writing user: %s metadata to cache engine.' % address)
+                    self.caching.set(address, response['user']['enterprise_user']['id'])
```

Comment:
I've tweaked the `cache_engine` references so they'll only invoke when `cache_engine == 'redis'` so that should maintain compatibility if someone is using a different caching solution; the Slack code just won't use it.